blockchain/utxoviewpoint.go

```diff
@@ -610,214 +503,185 @@ func (b *BlockChain) disconnectTransactions(view *UtxoViewpoint, block *dcrutil.
+	err := view.fetchInputUtxos(db, block)
+	if err != nil {
+		return err
+	}
+
+	// Connect all of the transactions in both the regular and stake trees of
+	// the block.  Notice that the stake tree is connected before the regular
+	// tree.  This means that stake transactions are not able to redeem outputs
+	// of transactions created in the regular tree of the same block, which is
+	// important since the regular tree may be disapproved by the subsequent
+	// block while the stake tree must remain valid.
+	for i, stx := range block.STransactions() {
+		err := view.connectTransaction(stx, block.Height(), uint32(i), stxos)
+		if err != nil {
+			return err
 		}
-		for txInIdx := len(msgTx.TxIn) - 1; txInIdx > -1; txInIdx-- {
-			// Ensure the spent txout index is decremented to stay
-			// in sync with the transaction input.
-			stxo := &stxos[stxoIdx]
-			stxoIdx--
+	}
+	for i, tx := range block.Transactions() {
+		err := view.connectTransaction(tx, block.Height(), uint32(i), stxos)
+		if err != nil {
+			return err
+		}
+	}
 
-			// When there is not already an entry for the referenced
-			// transaction in the view, it means it was fully spent,
-			// so create a new utxo entry in order to resurrect it.
-			txIn := msgTx.TxIn[txInIdx]
-			originHash := &txIn.PreviousOutPoint.Hash
-			originInIndex := txIn.PreviousOutPoint.Index
-			//originHeight := txIn.BlockHeight
-			// originIndex := txIn.BlockIndex
-			entry := view.entries[*originHash]
-			if entry == nil {
-				entry = newUtxoEntry(stxo.txVersion, stxo.height,
-					stxo.index, stxo.isCoinBase, stxo.hasExpiry,
-					stxo.txType)
-				if txType == stake.TxTypeSStx {
-					//stakeExtra := make([]byte,
-					//	serializeSizeForMinimalOutputs(tx))
-					// putTxToMinimalOutputs(stakeExtra, tx)
-					// entry.stakeExtra = stakeExtra
-					entry.stakeExtra = stxo.stakeExtra
-				}
-				view.entries[*originHash] = entry
-			}
+	// Update the best hash for view to include this block since all of its
+	// transactions have been connected.
+	view.SetBestHash(block.Hash())
+	return nil
+}
 
-			// Mark the entry as modified since it is either new
-			// or will be changed below.
-			entry.modified = true
+// disconnectBlock updates the view by disconnecting all of transactions in both
+// the regular and stake trees of the passed block, in the case the block
+// disapproves the parent block, reconnecting all of the transactions in the
+// regular tree of the previous block, and setting the best hash for the view to
+// the parent block.
+//
+// Connecting a transaction entails marking all utxos it spends as spent, and
+// adding all of the new utxos created by it.
+//
+// Disconnecting a transaction entails removing the utxos created by it and
+// restoring the outputs spent by it with the help of the provided spent txo
+// information.
+func (view *UtxoViewpoint) disconnectBlock(db database.DB, block, parent *dcrutil.Block, stxos []spentTxOut) error {
```

Comment:
Given that the usage of `stxos` is asymmetric between `connectBlock()` (where it's optional) and `disconnectBlock()` (where it's required) maybe we want to explicitly call that out (either here or in connectBlock).

Or change connectBlock to verify whether stxos is provided, given that is in fact needed for correct usage of the utxoviewpoint in the face of possibility of reorgs.